web/files/file.go

```diff
@@ -0,0 +1,151 @@
+package files
+
+import (
+	"bytes"
+	"crypto/md5" // #nosec
+	"encoding/json"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/couchdb"
+	"github.com/cozy/cozy-stack/web/jsonapi"
+	"github.com/spf13/afero"
+)
+
+type fileAttributes struct {
+	Name       string    `json:"name"`
+	CreatedAt  time.Time `json:"created_at"`
+	UpdatedAt  time.Time `json:"updated_at"`
+	Size       int64     `json:"size,string"`
+	Tags       []string  `json:"tags"`
+	MD5Sum     []byte    `json:"md5sum"`
+	Executable bool      `json:"executable"`
+	Class      string    `json:"class"`
+	Mime       string    `json:"mime"`
+}
+
+type fileDoc struct {
+	QID      string          `json:"_id"`
+	FRev     string          `json:"_rev,omitempty"`
+	Attrs    *fileAttributes `json:"attributes"`
+	FolderID string          `json:"folderID"`
+	Path     string          `json:"path"`
+}
+
+func (f *fileDoc) ID() string {
+	return f.QID
+}
+
+func (f *fileDoc) Rev() string {
+	return f.FRev
+}
+
+func (f *fileDoc) DocType() string {
+	return string(FileDocType)
+}
+
+func (f *fileDoc) SetID(id string) {
+	f.QID = id
+}
+
+func (f *fileDoc) SetRev(rev string) {
+	f.FRev = rev
+}
+
+// implement temporary interface JSONApier
+func (f *fileDoc) ToJSONApi() ([]byte, error) {
+	qid := f.QID
+	dat := map[string]interface{}{
+		"id":         qid[strings.Index(qid, "/")+1:],
+		"type":       f.DocType(),
+		"rev":        f.Rev(),
+		"attributes": f.Attrs,
+	}
+	m := map[string]interface{}{
+		"data": dat,
+	}
+	return json.Marshal(m)
+}
+
+// CreateFileAndUpload is the method for uploading a file onto the filesystem.
+func CreateFileAndUpload(m *DocMetadata, fs afero.Fs, dbPrefix string, body io.ReadCloser) (jsonapier jsonapi.JSONApier, err error) {
+	if m.Type != FileDocType {
+		err = errDocTypeInvalid
+		return
+	}
+
+	pth, _, err := createNewFilePath(m, fs, dbPrefix)
+	if err != nil {
+		return
+	}
+
+	createDate := time.Now()
+	attrs := &fileAttributes{
+		Name:       m.Name,
+		CreatedAt:  createDate,
+		UpdatedAt:  createDate,
+		Size:       int64(0),
+		Tags:       m.Tags,
+		MD5Sum:     m.GivenMD5,
+		Executable: m.Executable,
+		Class:      "document",   // @TODO
+		Mime:       "text/plain", // @TODO
+	}
+
+	doc := &fileDoc{
+		Attrs:    attrs,
+		FolderID: m.FolderID,
+		Path:     pth,
+	}
+
+	// Error handling to make sure the steps of uploading the file and
+	// creating the corresponding are both rollbacked in case of an
+	// error. This should preserve our VFS coherency a little.
+	defer func() {
+		if err == nil {
+			return
+		}
+		_, isCouchErr := err.(*couchdb.Error)
+		if isCouchErr {
+			couchdb.DeleteDoc(dbPrefix, doc)
+		} else {
+			fs.Remove(pth)
+		}
+	}()
```

Comment:
This error handler looks weird to me. I'd have simply use that:

```go
defer func() {
  if err != nil {
    fs.Remove(pth)
  }
}
```

What am I missing?